pkgs/tools/typesetting/tex/texlive/build-texlive-package.nix

```diff
@@ -202,5 +219,51 @@ let
       mkdir -p "$out"/share
       ln -s {"$texdoc"/doc,"$out"/share}/info
     '';
+
+  # most packages can be used as build inputs by adding their path to TEXMFAUXTREES
+  # exceptions are:
+  # - format binaries, which must be provided by a combined build input
+  # - hyphenation patterns, which cause formats to be rebuilt
+  # - occasional packages that look up in TEXMFROOT or similar
+
+  hasEnabledFormats = p: lib.any (f: f.enable or true) p.formats or [ ];
+
+  # resolve transitive dependencies instead of chaining propagated inputs for
+  # performance and because TeX Live has cyclic dependencies
+  # note that within TeX Live, it is safe to assume that distinct packages have
+  # distinct pname's
+  transitiveDeps = let key = p: { key = p.pname; inherit p; }; in
+    lib.catAttrs "p" (builtins.genericClosure {
+      startSet = [{ key = pname; p = tex; }];
+      # if a format is a transitive dependency, we may assume it will be
+      # provided by another combined build input
+      operator = item: builtins.map key (lib.filter (p: ! hasEnabledFormats p) item.p.tlDeps or [ ]);
+    });
+
+  # emit error if dev is evaluated on packages that cannot be used as inputs
+  # for clarity, emit error also if the package itself provides formats
+  disallowedAsBuildInput = hasEnabledFormats args
+    || lib.any (p: ! p.allowedAsBuildInput or true || p.hasHyphens or false) transitiveDeps;
+
+  # add tex folders to TEXMFAUXTREES, OSFONTDIR (for luaotfload)
+  dev = let container =
+    if disallowedAsBuildInput then
+      builtins.throw "The TeX Live package '${pname}' and/or its dependencies cannot be used as a build input. Please add '${pname}' to texlive.combine instead."
```

Comment:
There might be valid reasons to use one of the disallowed packages as a build input, for example if one wants to use a tex engine (without a preloaded format), so failing eval for this might be a little harsh? 

On the other side, such valid uses should be extremele rare, and the restriction could still be removed once such a use case emerges.

Reply:
True, one can compile formats on the side, like in the HOME folder. It's a dangerous use case (once you compile formats on your own, things get confusing) but we should make that configurable.

Reply:
> for example if one wants to use a tex engine (without a preloaded format), so failing eval for this might be a little harsh?

Expanding on this: if you launch `pdftex` from `texlive.bin.core`, it will try calling `mktexfmt` in user mode (i.e. output in `~/.texlive2022/...`), which then reads `fmtutil.cnf` and fails because `pdftex` is disabled. This is without getting into the `texlinks` side of the equation.

The solution would be an environment hook to `texlive.combine` that inspects the other build inputs for engines, generates a custom `fmtutil.cnf`, and runs `texlinks` on a temporary folder added in PATH... too much work I suspect!

PS: actually, one can override the error by specifying the output `texlive.__pkgs.FORMAT.out`.

Reply:
A related idea: using TeX Live as a build input in a shell has the side effect of adding potential garbage to `~/.texlive2022`. Even looking up a font means it will be added permanently to the database. How about setting `TEXMFVAR` to a temporary folder? It doesn't change anything for builds, it protects `HOME` when running a shell. The cost is the caches being lost when rerunning `nix-shell` (unless there is some deterministic temp folder we can use).